[Import/Export] Exported databases can now be safely moved

Tishj opened the pull request:
This PR fixes #5868 

Previously we would EXPORT and bake the absolute path of the table data into the query e.g `path/to/exported/db/tbl.csv`
When we only really need the `tbl.csv` part, because the path of the `db` folder is already provided on IMPORT.

Now we prefix the path with a placeholder on EXPORT, and replace it with the `db` folder path on IMPORT.

Tishj committed:
export now prefixes the filename with a placeholder, import replaces that placeholder with the path given to it

Tishj committed:
add test for import/export database with moving contents inbetween

Tishj commented:
Actually, do we even need the placeholder?

Mytherin commented:
Thanks for the PR!

While the placeholder approach is relatively straightforward I do see some potential issues that with this:

* It won't work correctly with older versions of DuckDB - as they won't know how to handle a placeholder, and won't emit a placeholder
* It won't work when loading into other systems, e.g. Postgres
* Global text replacement is a bit hacky, since the `IMPORT` file can contain arbitrary user-defined input (e.g. the actual file name)

Perhaps we can use the Parser for this - for example we can run the parser on the `load.sql` file. It should only contain `COPY` statements - and we can directly extract and modify the path for each of the `CopyStatements` within the file. We could then implement `CopyStatement::ToString` (similar to what we have for `SelectStatement`) to stringify it again. That should work while leaving the export code untouched.

Tishj commented:
You mean from new version -> old version
Is that supported in general?

Because for old version -> new version this should still work
it doesn't assert that the file is prefixed with the placeholder, only when it does it replaces it

The other points are fair though
I like the approach you suggested, I felt the string replacements were kinda dirty/error-prone to begin with

Mytherin commented:
> You mean from new version -> old version
> Is that supported in general?

It used to work, so I would prefer not to break it unnecessarily

Tishj committed:
parse the pragma, then change the file_path, stringify it again and use that

Tishj committed:
fixed issues

Tishj commented:
I've made some changes to the statement verifier to include the copy statement

Though I went a little overboard and implemented Equals for every Statement
I think I'll make a separate PR out of it, and also add Copy + Equals verification for all statements

Found and fixed a few bugs in this ToString that I need to move over to this PR, but other than that it should be done

Mytherin commented on src/parser/statement/copy_statement.cpp:
```diff
@@ -11,6 +11,101 @@ CopyStatement::CopyStatement(const CopyStatement &other) : SQLStatement(other),
+}
+
+string CopyStatement::ToString() const {
+	string result;
```
We have verification for `ToString` round-trips for INSERT/DELETE/SELECT statements in `src/main/client_context.cpp:670` (in `PendingStatementOrPreparedStatementInternal`). Perhaps add the COPY statement there as well?

Tishj committed:
add copy statement to ToString verification

Tishj committed:
take FORMAT into account in ToString, also options values is allowed to be empty (for example: HEADER)

Tishj committed:
Merge branch 'master' into absolute_import_export

Mytherin commented:
Thanks!
